### PR TITLE
Consolidate ISimpleFrameSite/IVBFormat/IGetVBAObject/DIPSID/IPropertyNotifySink

### PR DESCRIPTION
**Pull request by hughbe**
## Proposed changes

- Consolidate ISimpleFrameSite
- Consolidate IVBFormat
- Consolidate IGetVBAObject
- Consolidate DIPSID_*
- Consolidate IPropertyNotifySink

## Customer Impact

- Better performance, more resilient to crashes

## Regression? 

- No

## Risk

- Minimal

## Test methodology 

- Existing tests

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/1729)

### CODE REVIEW

File: src/Common/src/Interop/Ole32/Interop.IGetVBAObject.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        [ComImport]
+        [Guid("91733A60-3F4C-101B-A3F6-00AA0034E4E9")]
+        [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+        public unsafe interface IGetVBAObject
+        {
+            [PreserveSig]
+            HRESULT GetObject(
+                Guid* riid,
+                [Out, MarshalAs(UnmanagedType.LPArray)] IVBFormat[] rval,
```
Comment (hughbe) [inline src/Common/src/Interop/Ole32/Interop.IGetVBAObject.cs]: @JeremyKuhne is there a better way to marshal `IVBFormat**` parameters that is not arrays. Also is the `[Out]` attribute needed and is the `[MarshalAs(UnmanagedType.LPArray)]` attribute needed?

Unfortunately we can't seem to do `IVBFormat *`

File: src/Common/src/Interop/Ole32/Interop.DispId.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.IO;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class Ole32
+    {
+        public const int DISPID_VALUE = unchecked((int)0x0);
```
Comment (zsd4yr) [inline src/Common/src/Interop/Ole32/Interop.DispId.cs]: I think I would prefer these as an int enum `DispatchID`. Please add a comment indicating these are used in automation interfaces in order to identify methods, properties, and arguments

Also not sure how @JeremyKuhne 	feels about this, but I think these would all look better in decimal form...

File: src/Common/src/Interop/Oleaut32/Interop.VarFormat.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    internal static partial class Oleaut32
+    {
+        [Flags]
+        public enum VarFormatFlags : uint
+        {
+            VAR_FORMAT_NOSUBSTITUTE = 0x020,
+        }
+
+        [DllImport(Libraries.Oleaut32, ExactSpelling = true)]
+        public static extern HRESULT VarFormat(IntPtr pvarIn, IntPtr pstrFormat, int iFirstDay, int iFirstWeek, VarFormatFlags dwFlags, ref IntPtr pbstrOut);
```
Comment (zsd4yr) [inline src/Common/src/Interop/Oleaut32/Interop.VarFormat.cs]: ifirstDay should be an enum cast to an int here, the only valid options are
```
public enum VarFormatDay : int
{
	SystemDefault = 0,
	Monday = 1,
	Tuesday = 2,
	Wednesday = 3,
	Thursday = 4,
	Friday = 5,
	Saturday = 6,
	Sunday = 7,
}
```

Comment (zsd4yr) [reply src/Common/src/Interop/Oleaut32/Interop.VarFormat.cs]: iFirstWeek also only has a few values

0 is the SystemDefault, 1 is the first week contains january 1st, 2 is the larger half (foru days) of the first week is in the current year, and 3 is the first week has seven days. Sort of confusing to me, but anyways https://docs.microsoft.com/en-us/windows/win32/api/oleauto/nf-oleauto-varformat

File: src/Common/src/Interop/Ole32/Interop.IGetVBAObject.cs (hunk 1 above)
Comment (JeremyKuhne) [reply src/Common/src/Interop/Ole32/Interop.IGetVBAObject.cs]: You can just do `out IVBFormat ppvObj`.

File: src/Common/src/Interop/Oleaut32/Interop.VarFormat.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+using System.Runtime.InteropServices;
+
+internal partial class Interop
+{
+    internal static partial class Oleaut32
+    {
+        [Flags]
+        public enum VarFormatFlags : uint
+        {
+            VAR_FORMAT_NOSUBSTITUTE = 0x020,
```
Comment (zsd4yr) [inline src/Common/src/Interop/Oleaut32/Interop.VarFormat.cs]: This can also be `VAR_CALENDAR_HIJRI`, which I believe is 0x008

File: src/Common/src/Interop/Ole32/Interop.IGetVBAObject.cs (hunk 1 above)
Comment (hughbe) [reply src/Common/src/Interop/Ole32/Interop.IGetVBAObject.cs]: Is there a way to not throw NRE if the 
Param is null? Or to check if the double pointer is null as we do now.

File: src/Common/src/NativeMethods.cs [hunk 5]
```diff
@@ -789,26 +784,26 @@ public const int
...
+        public static uint MAKELCID(uint lgid)
         {
             return MAKELCID(lgid, SORT_DEFAULT);
         }
 
         /// <summary>
         ///  Creates an LCID from a LangId
         /// </summary>
-        public static int MAKELCID(int lgid, int sort)
+        public static uint MAKELCID(uint lgid, uint sort)
```
Comment (zsd4yr) [inline src/Common/src/NativeMethods.cs]: I think the first argument should be a MAKELANGID, maybe these changes should be in a different PR tho https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-makelangid

File: src/Common/src/Interop/Ole32/Interop.DispId.cs (hunk 2 above)
Comment (JeremyKuhne) [reply src/Common/src/Interop/Ole32/Interop.DispId.cs]: I'm starting to lean towards enums like this:

``` C#
public enum DISPID
{
    VALUE = 0,
    UNKNOWN = -1
    // etc...
}
```

The prefixes were done as there were no namespaces / named enums in C. It's cumbersome and hard to read when you're doing `DispatchId.DISPID_UNKNOWN` as opposed to `DISPID.UNKNOWN` (particularly with longer argument lists). It isn't 1-1 character wise with the C headers, but it's pretty close and isn't much of a learning curve given how many places we'd be doing this.

Having them sitting as consts in the namespace avoids the lengthier syntax, but starts to get a little cumbersome from a namespace pollution standpoint. If they're ever used as arguments it can make a huge productivity difference to have an enum when you're debugging.

File: src/Common/src/Interop/Ole32/Interop.IGetVBAObject.cs (hunk 1 above)
Comment (JeremyKuhne) [reply src/Common/src/Interop/Ole32/Interop.IGetVBAObject.cs]: _If_ there is a risk that this comes back null you can use `void**` and then get the interface via the `Marshal.GetObjectForIUnknown()` method. Something like this:

``` C#
void** rval;
// Call method & check HRESULT
if (rval != null && *rval != null)
{
    IVBFormat vbFormat = (IVBFormat)Marshal.GetObjectForIUnknown((IntPtr)*rval);
}
```

File: src/Common/src/Interop/Oleaut32/Interop.VarFormat.cs (hunk 3 above)
Comment (JeremyKuhne) [inline src/Common/src/Interop/Oleaut32/Interop.VarFormat.cs]: nit: this is getting long, can you break the arguments into separate lines?

File: src/Common/src/NativeMethods.cs [hunk 6]
```diff
@@ -4632,7 +4562,7 @@ public interface IProvideMultipleClassInfo
             // since the inheritance doesn't seem to work...
             // these are from IProvideClassInfo & IProvideClassInfo2
             [PreserveSig]
-            UnsafeNativeMethods.ITypeInfo GetClassInfo();
+            Interop.HRESULT GetClassInfo(out UnsafeNativeMethods.ITypeInfo ppTI);
```
Comment (JeremyKuhne) [inline src/Common/src/NativeMethods.cs]: Presumably we never called this particular method as it would choke on the prior interface we can just stub out the method.

For now that means `void** ppTI`, but ultimately this should be something like this:

``` C#
[PreserveSig]
void GetClassInfo_Stub();
// HRESULT GetClassInfo(out ITypeInfo ppTI);
```

The idea here is that we don't have RCW created by the interop layer generating unnecessary glue. It also makes it super explicit as to what we're actually using. Signatures do not matter if you don't call them, all that matters is that they are appropriately positioned in the interface definition. This pattern is given in ".NET and COM - The Complete Interoperability Guide", I tweak it to use the actual method names for extra clarity.

File: src/Common/src/NativeMethods.cs [hunk 7]
```diff
@@ -4666,8 +4596,8 @@ public class EVENTMSG
         [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
         public interface IProvideClassInfo
         {
-            [return: MarshalAs(UnmanagedType.Interface)]
-            UnsafeNativeMethods.ITypeInfo GetClassInfo();
+            [PreserveSig]
+            Interop.HRESULT GetClassInfo(out UnsafeNativeMethods.ITypeInfo ppTI);
```
Comment (JeremyKuhne) [inline src/Common/src/NativeMethods.cs]: I presume we never called this method, if so this whole interface could go and we can use `IntPtr` or `void*` where ever this interface was used.

File: src/Common/src/SafeNativeMethods.cs [hunk 8]
```diff
@@ -93,7 +90,7 @@ internal static class SafeNativeMethods
         public static extern bool PatBlt(HandleRef hdc, int left, int top, int width, int height, int rop);
 
         [DllImport(ExternDll.Kernel32, EntryPoint = "GetThreadLocale", CharSet = CharSet.Auto)]
-        public static extern int GetThreadLCID();
+        public static extern uint GetThreadLCID();
```
Comment (JeremyKuhne) [inline src/Common/src/SafeNativeMethods.cs]: If we're changing this we should probably change it to `GetThreadLocale` and get rid of the meaningless `CharSet`.

File: src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs [hunk 9]
```diff
@@ -5408,48 +5408,52 @@ protected void RaiseOnMouseDown(short button, short shift, int x, int y)
             base.OnMouseDown(new MouseEventArgs((MouseButtons)(((int)button) << 20), 1, x, y, 0));
         }
 
-        private class VBFormat : UnsafeNativeMethods.IVBFormat
+        private class VBFormat : Ole32.IVBFormat
         {
             // IVBFormat methods:
-            //
-            int UnsafeNativeMethods.IVBFormat.Format(ref object var, IntPtr pszFormat, IntPtr lpBuffer, short cpBuffer, int lcid, short firstD, short firstW, short[] result)
+            unsafe HRESULT Ole32.IVBFormat.Format(IntPtr vData, IntPtr bstrFormat, IntPtr lpBuffer, ushort cb, int lcid, short sFirstDayOfWeek, ushort sFirstWeekOfYear, ushort* rcb)
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs]: nit: can we break this one into separate lines as well?

File: src/Common/src/NativeMethods.cs (hunk 7 above)
Comment (hughbe) [reply src/Common/src/NativeMethods.cs]: It is used in `COM2TypeInfoProcessor.cs`

```cs
public static UnsafeNativeMethods.ITypeInfo FindTypeInfo(object obj, bool wantCoClass)
{
    UnsafeNativeMethods.ITypeInfo pTypeInfo = null;

    // This is kind of odd.  What's going on here is that if we want the CoClass (e.g. for
    // the interface name), we need to look for IProvideClassInfo first, then look for the
    // typeinfo from the IDispatch. In the case of many OleAut32 operations, the CoClass
    // doesn't have the interface members on it, although in the shell it usually does, so
    // we need to re-order the lookup if we *actually* want the CoClass if it's available.

    for (int i = 0; pTypeInfo == null && i < 2; i++)
    {
        if (wantCoClass == (i == 0))
        {
            if (obj is NativeMethods.IProvideClassInfo pProvideClassInfo)
            {
                try
                {
                    pProvideClassInfo.GetClassInfo(out pTypeInfo);
                }
                catch
                {
                }
            }
        }
        else
        {
            if (obj is UnsafeNativeMethods.IDispatch iDispatch)
            {
                try
                {
                    iDispatch.GetTypeInfo(0,Kernel32.GetThreadLocale(), out pTypeInfo);
                }
                catch
                {
                }
            }
        }
    }

    return pTypeInfo;
}
```

File: src/Common/src/Interop/Interop.HRESULT.cs [hunk 10]
```diff
@@ -22,4 +25,6 @@ internal enum HRESULT : int
         E_ACCESSDENIED = unchecked((int)0x80070005L),
         E_INVALIDARG = unchecked((int)0x80070057),
     }
+
+    public static bool Succeeded(HRESULT hr) => hr >= 0;
```
Comment (zsd4yr) [inline src/Common/src/Interop/Interop.HRESULT.cs]: I would add this as an extension method on the enum instead 

https://stackoverflow.com/questions/15388072/how-to-add-extension-methods-to-enums/37634795

something like:

```c#
public static class HRESULTExtensions 
{
	public static bool Succeeded(this HRESULT hresult) 
		=> hr >= 0;
}
```

And then when you use it, you should be able to just

```c#
HRESULT hr = ...;
bool result = hr.Succeeded();
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/COM2AboutBoxPropertyDescriptor.cs [hunk 11]
```diff
@@ -14,7 +15,7 @@ internal class Com2AboutBoxPropertyDescriptor : Com2PropertyDescriptor
         private TypeConverter converter;
         private UITypeEditor editor;
 
-        public Com2AboutBoxPropertyDescriptor() : base(NativeMethods.ActiveX.DISPID_ABOUTBOX, "About", new Attribute[]{new DispIdAttribute(NativeMethods.ActiveX.DISPID_ABOUTBOX),
+        public Com2AboutBoxPropertyDescriptor() : base(Ole32.DISPID_ABOUTBOX, "About", new Attribute[]{new DispIdAttribute(Ole32.DISPID_ABOUTBOX),
```
Comment (zsd4yr) [inline src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/COM2AboutBoxPropertyDescriptor.cs]: I really feel like all these strings "About", etc... should be defined as consts somewhere or else you should use nameof(About) if there is some reasonable way to do that.

Comment (zsd4yr) [inline src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/COM2AboutBoxPropertyDescriptor.cs]: When you make all of those disp IDs into an enum, see if you can change the DispIdAttribute to take one of those instead of an int

Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/COM2AboutBoxPropertyDescriptor.cs]: Its a CLR public API that we can't extend. Casting will be needed in `DispIdAttribute`

File: src/Common/src/NativeMethods.cs (hunk 7 above)
Comment (JeremyKuhne) [reply src/Common/src/NativeMethods.cs]: My mistake, I didn't see that you changed it to `[PreserveSig]`. You should either change the signature back and remove the `[return]` or check the HRESULT and remove the `try..catch` blocks. Since we're ignoring everything I'd prefer to remove the catch all blocks- since generally they're evil.

File: src/Common/src/Interop/Interop.HRESULT.cs (hunk 10 above)
Comment (hughbe) [reply src/Common/src/Interop/Interop.HRESULT.cs]: I don't really like it as there's a whole load of areas where we write `if (Succeeded(...))` and this matches the style used in c++/c which reads more clearly in interop scenarios

File: src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/COM2AboutBoxPropertyDescriptor.cs (hunk 11 above)
Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/COM2AboutBoxPropertyDescriptor.cs]: I think its best to do constant extraction in another PR

File: src/Common/src/Interop/Interop.HRESULT.cs (hunk 10 above)
Comment (JeremyKuhne) [reply src/Common/src/Interop/Interop.HRESULT.cs]: It is what we're already doing in the designer codebase, so we'll be doing it here eventually as well. While it doesn't match C++ syntax it helps with discovery, which I think is worth the difference.

``` C#
using static Interop;

internal static partial class Interop
{
    public enum HRESULT : int
    {
        S_OK = 0,
        S_FALSE = 1,
        E_NOTIMPL = unchecked((int)0x80004001),
        E_NOINTERFACE = unchecked((int)0x80004002),
        // ...
    }
}

internal static class HResultExtensions
{
    public static bool Succeeded(this HRESULT hr) => hr >= 0;

    public static bool Failed(this HRESULT hr) => hr < 0;
}
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/COM2AboutBoxPropertyDescriptor.cs (hunk 11 above)
Comment (weltkante) [reply src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/COM2AboutBoxPropertyDescriptor.cs]: Also I think DISPID are per-interface so there can be lots of conflicts, which might be an argument not to make it an enum ...

File: src/Common/src/Interop/Interop.HRESULT.cs (hunk 10 above)
Comment (zsd4yr) [reply src/Common/src/Interop/Interop.HRESULT.cs]: I would also argue that it should not look "c++-ish", it should look c#-ish

But, yeah, prior art -- thanks Jeremy

File: src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/COM2AboutBoxPropertyDescriptor.cs (hunk 11 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/ComponentModel/COM2Interop/COM2AboutBoxPropertyDescriptor.cs]: > there can be lots of conflicts, which might be an argument not to make it an enum

Yeah, in that case a static class would be better.
